File: advent-of-code/2021/day2-2.py
def solve(instructions):
    x = 0
    y = 0
    aim = 0

    for direction, magnitude in instructions:
        magnitude = int(magnitude)
        if direction == "forward":
            x += magnitude
            y += aim * magnitude
        if direction == "down":
            aim += magnitude
        if direction == "up":
            aim -= magnitude

    return x * y

def main():
    with open("input/2.data") as f:
        instructions = list(map(lambda line: line.strip().split(), f.readlines()))
        print(solve(instructions))

if __name__ == "__main__":
    main()

File: advent-of-code/2021/day2-2.hs
processCommand :: [Int] -> [String] -> [Int]
processCommand [x, y, aim] [dir, mag]
  | dir == "forward" = [x + mag_, y + aim * mag_, aim]
  | dir == "up"      = [x, y, aim - mag_]
  | dir == "down"    = [x, y, aim + mag_]
  where
    mag_ = read mag :: Int

solve :: [String] -> Int
solve commands = x * y
  where
    [x, y, _] = foldl processCommand [0, 0, 0] (map words commands)

main :: IO ()
main = do
  input <- readFile "input/2.data"
  putStrLn $ (show . solve) (lines input)
